aptos-move/framework/aptos-framework/sources/Marketplace.move

```diff
@@ -0,0 +1,289 @@
+// This module provides a basic NFT Marketplace with first-price auction and fixed-price sale
+module AptosFramework::Marketplace {
+    use AptosFramework::Coin::{Self, Coin};
+    use AptosFramework::Table::{Self, Table};
+    use AptosFramework::TestCoin::TestCoin;
+    use AptosFramework::Timestamp;
+    use AptosFramework::Token::{Self, Token, TokenId};
+    use Std::Signer;
+    use Std::Event::{Self, EventHandle};
+
+    const ERROR_INVALID_BUYER: u64 = 0;
+    const ERROR_INSUFFICIENT_BID: u64 = 1;
+    const ERROR_AUCTION_INACTIVE: u64 = 2;
+    const ERROR_AUCTION_NOT_COMPLETE: u64 = 3;
+    const ERROR_NOT_CLAIMABLE: u64 = 4;
+    const ERROR_CLAIM_COINS_FIRST: u64 = 5;
+
+    struct AuctionItem has store, drop {
+        min_bid: u64,
+        duration: u64,
+        start_time: u64,
+        curr_bid: u64,
+        curr_bidder: address,
+    }
+
+    struct ListedItem has store, drop {
+        price: u64
+    }
+
+    struct CoinEscrow has key {
+        locked_coins: Table<TokenId, Coin<TestCoin>>,
+    }
+
+    struct TokenEscrow has key {
+        locked_tokens: Table<TokenId, Token>,
+    }
+
+    // Set of data sent to the event stream during a auctioning a token
+    struct AuctionEvent has drop, store {
+        id: TokenId,
+        min_bid: u64,
+        duration: u64
+    }
+
+    // Set of data sent to the event stream during a bidding for a token
+    struct BidEvent has drop, store {
+        id: TokenId,
+        bid: u64,
+    }
+
+    // Set of data sent to the event stream during a bidding for a token
+    struct ClaimCoinsEvent has drop, store {
+        id: TokenId,
+    }
+
+    // Set of data sent to the event stream during a bidding for a token
+    struct ClaimTokenEvent has drop, store {
+        id: TokenId,
+    }
+
+    // Set of data sent to the event stream during a listing of a token (for fixed price)
+    struct ListEvent has drop, store {
+        id: TokenId,
+        amount: u64,
+    }
+
+    // Set of data sent to the event stream during a buying of a token (for fixed price)
+    struct BuyEvent has drop, store {
+        id: TokenId,
+    }
+
+    struct AuctionData has key {
+        auction_items: Table<TokenId, AuctionItem>,
+        auction_events: EventHandle<AuctionEvent>,
+        bid_events: EventHandle<BidEvent>,
+        claim_coins_events: EventHandle<ClaimCoinsEvent>,
+        claim_token_events: EventHandle<ClaimTokenEvent>,
+    }
+
+    struct ListedItemsData has key {
+        listed_items: Table<TokenId, ListedItem>,
+        listing_events: EventHandle<ListEvent>,
+        buying_events: EventHandle<BuyEvent>,
+    }
+
+    public(script)fun initialize_auction(account: &signer, creator: address, collection_name: vector<u8>, name: vector<u8>, min_bid: u64, duration: u64) acquires TokenEscrow, AuctionData {
```

Comment:
should we only allow token owners to init auction? should we also allow 3rd to run auction on our behalf?